src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQConnection.java

```diff
@@ -207,20 +242,92 @@ public void publish(JSONObject json) {
      * Sends messages from the message queue.
      */
     private void sendMessages() {
+        Channel channel = null;
+
         while (true) {
             try {
-                MessageData messageData = (MessageData)messageQueue.poll(SENDMESSAGE_TIMEOUT,
+                if (channel == null || !channel.isOpen()) {
+                    channel = createChannel();
+                    channel.confirmSelect();
+                    addMessageConfirmListener(channel);
+                }
+                MessageData messageData = (MessageData) messageQueue.poll(SENDMESSAGE_TIMEOUT,
                                                                          TimeUnit.MILLISECONDS);
                 if (messageData != null) {
-                    getInstance().send(messageData.getExchange(), messageData.getRoutingKey(),
-                            messageData.getProps(), messageData.getBody());
+                    validateExchange(channel, messageData.getExchange());
+                    getInstance().sendOnChannel(messageData, channel);
                 }
             } catch (InterruptedException ie) {
                 LOGGER.info("sendMessages() poll() was interrupted: ", ie);
+            } catch (IOException | NullPointerException ioe) {
+                LOGGER.error("error validating channel: ", ioe);
+            } catch (ChannelCreationException | MessageDeliveryException transientException) {
+                LOGGER.error(transientException.getMessage(), transientException.getCause());
+                try {
+                    Thread.sleep(CONNECTION_WAIT);
+                } catch (InterruptedException ie) {
+                    LOGGER.error("Thread.sleep() was interrupted", ie);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Validate the exchange.
+     *
+     * @throws IOException if the channel is invalid for the exchange
+     */
+    private void validateExchange(Channel channel, String exchange) throws IOException, NullPointerException {
```

Comment:
Can this throw NPE?

Reply:
Mistake from my part, it should have thrown a NPE if exchange  == null. I've changed it to IllegalArgumentException instead.